pkg/core/blockchain.go

```diff
@@ -679,9 +716,55 @@ func (bc *Blockchain) GetStateModule() blockchainer.StateRoot {
 	return bc.stateRoot
 }
 
+// saveBlock stores specified block skipping executable scripts and updates
+// Blockchain state.
+func (bc *Blockchain) saveBlock(block *block.Block) error {
+	cache := dao.NewCached(bc.dao)
+	writeBuf := io.NewBufBinWriter()
+	if err := cache.StoreAsBlock(block, writeBuf); err != nil {
+		return err
+	}
+	writeBuf.Reset()
+
+	if err := cache.StoreAsCurrentBlock(block, writeBuf); err != nil {
+		return err
+	}
+	writeBuf.Reset()
+
+	for _, tx := range block.Transactions {
+		if err := cache.StoreAsTransaction(tx, block.Index, writeBuf); err != nil {
+			return err
+		}
+		writeBuf.Reset()
+
+		err := bc.storeConflictTxs(tx, block, writeBuf, cache)
+		if err != nil {
+			return err
+		}
+	}
+	if bc.config.SaveStorageBatch {
+		bc.lastBatch = cache.DAO.GetBatch()
+	}
+
+	bc.lock.Lock()
+	_, err := cache.Persist()
+	if err != nil {
+		bc.lock.Unlock()
+		return err
+	}
+	bc.topBlock.Store(block)
+	atomic.StoreUint32(&bc.blockHeight, block.Index)
+	bc.lock.Unlock()
+
+	updateBlockHeightMetric(block.Index)
+
+	// need to notify server when last block is stored
+	bc.events <- bcEvent{block, nil}
```

Comment:
Do we really need it? It's not processed properly.

Reply:
We need this notification only for the `p1`-th block to notify the server that block sync is reached. That seemed to me the most reliable way to do it.